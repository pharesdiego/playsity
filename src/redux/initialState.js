const initialState = {
  videoplayer: {
    url: '/video.mp4'
  },
  playlist: {
    clips: [
      {
        id: null,
        name: 'Full Video',
        start: '0',
        end: '0',
        tags: ''
      }
    ],
    currentClip: null
  }
};

export default initialState;