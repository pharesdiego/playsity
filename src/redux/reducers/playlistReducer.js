import {
  PLAYLIST_CLIP_CREATE,
  PLAYLIST_CLIP_EDIT,
  PLAYLIST_CURRENT_CLIP_UPDATE,
  PLAYLIST_CLIP_DELETE
} from './actions';

import initialState from './../initialState';
import key from './../../utils/key';
import rangeValidator from './../../utils/rangeValidator';

const appState = (JSON.parse(window.localStorage.getItem('PLAYSITY_STORE')) || initialState).playlist;

const playlistReducer = (state = appState, action) => {
  switch(action.type) {
    case PLAYLIST_CLIP_CREATE:
      let { start, end, name, tags, TEST_ID } = action;
      let id = TEST_ID || key();
      let [min, max] = rangeValidator(+start, +end);
      return {
        ...state,
        clips: [
          ...state.clips, 
          {
            id,
            name: name || id,
            start: (min+''),
            end: (max+''),
            tags: tags || ''
          }
        ]
      }
    case PLAYLIST_CLIP_EDIT:
      return {
        ...state,
        clips: [
          ...editClip(state, action)
        ]
      }
    case PLAYLIST_CLIP_DELETE:
      return {
        ...state,
        clips: [
          ...state.clips.filter(clip => action.id !== clip.id)
        ]
      }

    case PLAYLIST_CURRENT_CLIP_UPDATE:
      return {
        ...state,
        currentClip: action.id
      }
    default:
      return state;
  }
}

const editClip = (state, { id, name, start, end, tags }) => {
  let [min, max] = rangeValidator(+start, +end);
  return state
          .clips
          .map(clip => clip.id === id ? (
            {
              id, 
              name: name || id, 
              start: (min+''), 
              end: (max+''), 
              tags 
            }
          ) : clip);
}

export default playlistReducer;