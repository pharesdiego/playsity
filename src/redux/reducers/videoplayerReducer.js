import {
  VIDEO_URL_UPDATE
} from './actions';

import initialState from './../initialState';

const appState = (JSON.parse(window.localStorage.getItem('store')) || initialState).videoplayer;

const playlistReducer = (state = appState, action) => {
  switch(action.type) {
    case VIDEO_URL_UPDATE: 
      return {
        ...state,
        url: action.url
      }
    default:
      return state;
  }
}

export default playlistReducer;