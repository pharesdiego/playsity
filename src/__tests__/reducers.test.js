import playlistReducer from './../redux/reducers/playlistReducer';
import videoplayerReducer from './../redux/reducers/videoplayerReducer';
import initialState from './../redux/initialState';
import {
  VIDEO_URL_UPDATE,
  PLAYLIST_CLIP_CREATE,
  PLAYLIST_CLIP_EDIT,
  PLAYLIST_CURRENT_CLIP_UPDATE,
  PLAYLIST_CLIP_DELETE
} from './../redux/reducers/actions';

const { playlist, videoplayer } = initialState;

describe('>>> Playlist Reducers <<<', () => {
  it('PLAYLIST_CLIP_CREATE', () => {
    expect(
      playlistReducer(playlist, {
        type: PLAYLIST_CLIP_CREATE,
        TEST_ID: '1',
        name: 'Final Scene',
        start: '40',
        end: '52',
        tags: 'final, credits'
      })
    ).toMatchSnapshot();

    expect(
      playlistReducer(playlist, {
        type: PLAYLIST_CLIP_CREATE,
        TEST_ID: '1',
        name: '',
        start: '',
        end: '',
        tags: ''
      })
    ).toMatchSnapshot();

    expect(
      playlistReducer(playlist, {
        type: PLAYLIST_CLIP_CREATE,
        TEST_ID: '1',
        name: '',
        start: '10000000000',
        end: '100000000000',
        tags: ''
      })
    ).toMatchSnapshot();
  });

  it('PLAYLIST_CLIP_EDIT', () => {
    expect(
      playlistReducer(playlist, {
        type: PLAYLIST_CLIP_EDIT,
        id: 'a21e19da2dc',
        name: 'Intro Edited',
        start: '1',
        end: '12',
        tags: 'Snow, Walking, Mountains, Running'
      })
    ).toMatchSnapshot();

    expect(
      playlistReducer(playlist, {
        type: PLAYLIST_CLIP_EDIT,
        id: 'a21e19da2dc',
        name: 'Time Edited',
        start: '10000000',
        end: '120000000000',
        tags: 'Snow, Walking, Mountains, Running'
      })
    ).toMatchSnapshot();
  });

  it('PLAYLIST_CURRENT_CLIP_UPDATE', () => {
    expect(
      playlistReducer(playlist, {
        type: PLAYLIST_CURRENT_CLIP_UPDATE,
        id: '9ea3abcc842'
      })
    ).toMatchSnapshot();

    expect(
      playlistReducer(playlist, {
        type: PLAYLIST_CURRENT_CLIP_UPDATE,
        id: 'changed current clip'
      })
    ).toMatchSnapshot();
  });

  it('PLAYLIST_CLIP_DELETE', () => {
    expect(
      playlistReducer(playlist, {
        type: PLAYLIST_CLIP_DELETE,
        id: '9ea3abcc842'
      })
    ).toMatchSnapshot();
  })
});

describe('>>> Videoplayer Reducers <<<', () => {
  it('VIDEO_URL_UPDATE', () => {
    expect(
      videoplayerReducer(videoplayer, {
        type: VIDEO_URL_UPDATE,
        start: '0',
        end: '15'
      })
    ).toMatchSnapshot();
  });
});