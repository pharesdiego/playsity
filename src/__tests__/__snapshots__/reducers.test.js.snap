// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> Playlist Reducers <<< PLAYLIST_CLIP_CREATE 1`] = `
Object {
  "clips": Array [
    Object {
      "end": "52",
      "id": null,
      "name": "Full Video",
      "start": "0",
      "tags": "",
    },
    Object {
      "end": "11",
      "id": "a21e19da2dc",
      "name": "Intro",
      "start": "0",
      "tags": "Snow, Walking, Mountains",
    },
    Object {
      "end": "45",
      "id": "9ea3abcc842",
      "name": "A dangerous quest",
      "start": "35",
      "tags": "",
    },
    Object {
      "end": "52",
      "id": "1",
      "name": "Final Scene",
      "start": "40",
      "tags": "final, credits",
    },
  ],
  "currentClip": null,
}
`;

exports[`>>> Playlist Reducers <<< PLAYLIST_CLIP_CREATE 2`] = `
Object {
  "clips": Array [
    Object {
      "end": "52",
      "id": null,
      "name": "Full Video",
      "start": "0",
      "tags": "",
    },
    Object {
      "end": "11",
      "id": "a21e19da2dc",
      "name": "Intro",
      "start": "0",
      "tags": "Snow, Walking, Mountains",
    },
    Object {
      "end": "45",
      "id": "9ea3abcc842",
      "name": "A dangerous quest",
      "start": "35",
      "tags": "",
    },
    Object {
      "end": "0",
      "id": "1",
      "name": "1",
      "start": "0",
      "tags": "",
    },
  ],
  "currentClip": null,
}
`;

exports[`>>> Playlist Reducers <<< PLAYLIST_CLIP_CREATE 3`] = `
Object {
  "clips": Array [
    Object {
      "end": "52",
      "id": null,
      "name": "Full Video",
      "start": "0",
      "tags": "",
    },
    Object {
      "end": "11",
      "id": "a21e19da2dc",
      "name": "Intro",
      "start": "0",
      "tags": "Snow, Walking, Mountains",
    },
    Object {
      "end": "45",
      "id": "9ea3abcc842",
      "name": "A dangerous quest",
      "start": "35",
      "tags": "",
    },
    Object {
      "end": "52",
      "id": "1",
      "name": "1",
      "start": "0",
      "tags": "",
    },
  ],
  "currentClip": null,
}
`;

exports[`>>> Playlist Reducers <<< PLAYLIST_CLIP_DELETE 1`] = `
Object {
  "clips": Array [
    Object {
      "end": "52",
      "id": null,
      "name": "Full Video",
      "start": "0",
      "tags": "",
    },
    Object {
      "end": "11",
      "id": "a21e19da2dc",
      "name": "Intro",
      "start": "0",
      "tags": "Snow, Walking, Mountains",
    },
  ],
  "currentClip": null,
}
`;

exports[`>>> Playlist Reducers <<< PLAYLIST_CLIP_EDIT 1`] = `
Object {
  "clips": Array [
    Object {
      "end": "52",
      "id": null,
      "name": "Full Video",
      "start": "0",
      "tags": "",
    },
    Object {
      "end": "12",
      "id": "a21e19da2dc",
      "name": "Intro Edited",
      "start": "1",
      "tags": "Snow, Walking, Mountains, Running",
    },
    Object {
      "end": "45",
      "id": "9ea3abcc842",
      "name": "A dangerous quest",
      "start": "35",
      "tags": "",
    },
  ],
  "currentClip": null,
}
`;

exports[`>>> Playlist Reducers <<< PLAYLIST_CLIP_EDIT 2`] = `
Object {
  "clips": Array [
    Object {
      "end": "52",
      "id": null,
      "name": "Full Video",
      "start": "0",
      "tags": "",
    },
    Object {
      "end": "52",
      "id": "a21e19da2dc",
      "name": "Time Edited",
      "start": "0",
      "tags": "Snow, Walking, Mountains, Running",
    },
    Object {
      "end": "45",
      "id": "9ea3abcc842",
      "name": "A dangerous quest",
      "start": "35",
      "tags": "",
    },
  ],
  "currentClip": null,
}
`;

exports[`>>> Playlist Reducers <<< PLAYLIST_CURRENT_CLIP_UPDATE 1`] = `
Object {
  "clips": Array [
    Object {
      "end": "52",
      "id": null,
      "name": "Full Video",
      "start": "0",
      "tags": "",
    },
    Object {
      "end": "11",
      "id": "a21e19da2dc",
      "name": "Intro",
      "start": "0",
      "tags": "Snow, Walking, Mountains",
    },
    Object {
      "end": "45",
      "id": "9ea3abcc842",
      "name": "A dangerous quest",
      "start": "35",
      "tags": "",
    },
  ],
  "currentClip": "9ea3abcc842",
}
`;

exports[`>>> Playlist Reducers <<< PLAYLIST_CURRENT_CLIP_UPDATE 2`] = `
Object {
  "clips": Array [
    Object {
      "end": "52",
      "id": null,
      "name": "Full Video",
      "start": "0",
      "tags": "",
    },
    Object {
      "end": "11",
      "id": "a21e19da2dc",
      "name": "Intro",
      "start": "0",
      "tags": "Snow, Walking, Mountains",
    },
    Object {
      "end": "45",
      "id": "9ea3abcc842",
      "name": "A dangerous quest",
      "start": "35",
      "tags": "",
    },
  ],
  "currentClip": "changed current clip",
}
`;

exports[`>>> Videoplayer Reducers <<< VIDEO_URL_UPDATE 1`] = `
Object {
  "url": "/video.mp4#t=0,15",
}
`;
